### PR TITLE
Change pointer arithmetic to use getelementptr.

### PR DESCRIPTION
**Pull request by rdomingu**
Change inttoptr ( add x, const) -> gep ( inttoptr x, const) to
improve value tracking and load/store vectorization.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit 4c91ff31f6894fee886cf68fbfa953867401b450
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 4ae736bdbc5d5dee59851ac564c5e21d807b44b0</li>
<li>SPVGEN: dev e1f1881d84cd1c7a6c1d661cbac0e619b4e110c1</li>
<li>XGL: dev 87430a6fa88da11473eed49603def2fa47b51fa0</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev c9860f1b279efa64b57714a17025e32d12dbb593</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/140859)</summary><ul><li>Built with version 1.2.2.2</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 26166/46960 (55.7%)</li>
<li>Failed: 0/46960 (0.0%)</li>
<li>Skipped: 20794/46960 (44.3%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 26165/46960 (55.7%)</li>
<li>Failed: 0/46960 (0.0%)</li>
<li>Skipped: 20795/46960 (44.3%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 24702/46939 (52.6%)</li>
<li>Failed: 0/46939 (0.0%)</li>
<li>Skipped: 22237/46939 (47.4%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 6578a7544e35d58c5366ef7c9c6cd3af864362d5
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 4ae736bdbc5d5dee59851ac564c5e21d807b44b0</li>
<li>SPVGEN: dev e1f1881d84cd1c7a6c1d661cbac0e619b4e110c1</li>
<li>XGL: dev 87430a6fa88da11473eed49603def2fa47b51fa0</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev c9860f1b279efa64b57714a17025e32d12dbb593</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/140859)</summary><ul><li>Built with version 1.2.2.2</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 26166/46960 (55.7%)</li>
<li>Failed: 0/46960 (0.0%)</li>
<li>Skipped: 20794/46960 (44.3%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 26165/46960 (55.7%)</li>
<li>Failed: 0/46960 (0.0%)</li>
<li>Skipped: 20795/46960 (44.3%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 24702/46939 (52.6%)</li>
<li>Failed: 0/46939 (0.0%)</li>
<li>Skipped: 22237/46939 (47.4%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 300ea88871a8ffc183f9bf42a8cf84f264637c6c
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 4ae736bdbc5d5dee59851ac564c5e21d807b44b0</li>
<li>SPVGEN: dev e1f1881d84cd1c7a6c1d661cbac0e619b4e110c1</li>
<li>XGL: dev 87430a6fa88da11473eed49603def2fa47b51fa0</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev c9860f1b279efa64b57714a17025e32d12dbb593</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/140859)</summary><ul><li>Built with version 1.2.2.2</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 26166/46960 (55.7%)</li>
<li>Failed: 0/46960 (0.0%)</li>
<li>Skipped: 20794/46960 (44.3%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 26165/46960 (55.7%)</li>
<li>Failed: 0/46960 (0.0%)</li>
<li>Skipped: 20795/46960 (44.3%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 24702/46939 (52.6%)</li>
<li>Failed: 0/46939 (0.0%)</li>
<li>Skipped: 22237/46939 (47.4%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 055b7e66f686739ffab3c6fd49423f9df8f20b39
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 4ae736bdbc5d5dee59851ac564c5e21d807b44b0</li>
<li>SPVGEN: dev e1f1881d84cd1c7a6c1d661cbac0e619b4e110c1</li>
<li>XGL: dev 87430a6fa88da11473eed49603def2fa47b51fa0</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev c9860f1b279efa64b57714a17025e32d12dbb593</li>
</ul></details>

<details><summary>CTS tests (Failed: 1/140859)</summary><ul><li>Built with version 1.2.2.2</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 26166/46960 (55.7%)</li>
<li>Failed: 0/46960 (0.0%)</li>
<li>Skipped: 20794/46960 (44.3%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 26164/46960 (55.7%)</li>
<li><details><summary>Failed: 1/46960 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.memory.pipeline_barrier.transfer_src_transfer_dst.1048576
Stack trace: Flake
```
</p></details></li>

<li>Skipped: 20795/46960 (44.3%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 24702/46939 (52.6%)</li>
<li>Failed: 0/46939 (0.0%)</li>
<li>Skipped: 22237/46939 (47.4%)</li>
</ul></details>


</ul></details>

**Review by kuhar — CHANGES_REQUESTED** [review]
My high level question is why this isn't handled by generic instcombine transformations? Is it because of some corner cases with ptr2int instructions that must be comparable, but geps wouldn't be? If so, what extra assumptions make this custom transform correct for shader code?

**Comment by rdomingu** [comment]
> My high level question is why this isn't handled by generic instcombine transformations? Is it because of some corner cases with ptr2int instructions that must be comparable, but geps wouldn't be? If so, what extra assumptions make this custom transform correct for shader code?

This transformation wouldn't be legal in LLVM. In principle, the pointer returned by inttoptr ( add x, const ) is different from the pointer returned by gep ( inttoptr x, const ): the latter would be associated with whatever x AND const were associated; the former would be associated ONLY with whatever x was associated. We can make the assumption in LLPC that const is not associated with any object but that's not assumption that can be made in general in LLVM (with all its different front-ends). See this related discussion: https://groups.google.com/g/llvm-dev/c/x4K7ppGLbg8/m/f_3NySRhjlcJ

**Comment by kuhar** [comment]
> > My high level question is why this isn't handled by generic instcombine transformations? Is it because of some corner cases with ptr2int instructions that must be comparable, but geps wouldn't be? If so, what extra assumptions make this custom transform correct for shader code?
> 
> This transformation wouldn't be legal in LLVM. In principle, the pointer returned by inttoptr ( add x, const ) is different from the pointer returned by gep ( inttoptr x, const ): the latter would be associated with whatever x AND const were associated; the former would be associated ONLY with whatever x was associated. We can make the assumption in LLPC that const is not associated with any object but that's not assumption that can be made in general in LLVM (with all its different front-ends). See this related discussion: https://groups.google.com/g/llvm-dev/c/x4K7ppGLbg8/m/f_3NySRhjlcJ

I think it's very important to clearly state things like this in code and in commit messages. Can you update both and also include the link to the llvm-dev discussion?

But back to the subject, I'm a bit lost in both the llvm-dev thread and your reply because it seems to involve alias analysis but uses a non-standard terminology. I assume that a value being 'associated' with an object means provenance tracking in this context, like explained in [1]?

IE when you have `%a = int2ptr(x + const)` and `%b = gep(int2ptr(x), const)`, `%a` doesn't have the same points-to set as `points-to(x)` because llvm has to assume it may cross over and access another objects, while `%b` has the points-to set as `points-to(x)` (under the assumption that `const` has an empty points-to set) and points to a different offset at the same object as `x`. And by optimizing `%a` into `%b`, we allow more optimizations because we can assume that `points-to(const)` is always an empty set? This seems to be the other way round in your explanation.

[1] https://www.cs.utah.edu/~regehr/oopsla18.pdf sections 2.2-2.4

**Comment by rdomingu** [comment]
You're right. I got it backwards. It should be: "This transformation wouldn't be legal in LLVM. In principle, the pointer returned by inttoptr ( add x, const ) is different from the pointer returned by gep ( inttoptr x, const ): the **former** would be associated with whatever x AND const were associated; the **latter** would be associated ONLY with whatever x was associated.

**Comment by kuhar** [comment]
But what makes it possible to make this assumption in the spirv-reader/lgc frontend?

I guess it we can consider these cases separately:
1. Local variables (function-level): illegal to do pointer arithmetic and cross over to another local variable on SPIR-V level (?)
2. Global buffers: illegal to do pointer arithmetic and go outside of these on the SPIR-V/PAL level (?)
3. Anything else?

File: lgc/patch/PatchPeepholeOpt.cpp [hunk 1]
```diff
@@ -571,6 +571,45 @@ void PatchPeepholeOpt::visitPHINode(PHINode &phiNode) {
   }
 }
 
+// =====================================================================================================================
+// Visit an inttoptr instruction.
+//
+// @param intToPtr: The "inttoptr" instruction to visit.
+void PatchPeepholeOpt::visitIntToPtr(IntToPtrInst &intToPtr) {
+  // Check if we are using add to do pointer arithmetic.
+  if (BinaryOperator *const binaryOperator = dyn_cast<BinaryOperator>(intToPtr.getOperand(0))) {
```
[the remark this answers is not in the document]
Comment (rdomingu) [reply lgc/patch/PatchPeepholeOpt.cpp:580]: Done.

**Comment by rdomingu** [comment]
"%a doesn't have the same points-to set as points-to(x) because llvm has to assume it may cross over and access another objects,"

The problem is not the pointer crossing over and accessing another object. The problem is that LLVM can't assume that x is the pointer and const is the offset. It could be the other way around: x could be the offset and const could be the pointer. So LLVM has to be conservative and make points-to(%a) = points-to(x) + points-to(const).

On the other hand, in LLPC (and SPIRV) we don't have any cases where const could be a pointer. Therefore, it is safe to make points-to(%a) = points-to(x).

**Comment by amdvlk-admin** [comment]
### Test summary for commit 8e95371554a60a7b8a09fc6271c56734fd0e78ff
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 4ae736bdbc5d5dee59851ac564c5e21d807b44b0</li>
<li>SPVGEN: dev e1f1881d84cd1c7a6c1d661cbac0e619b4e110c1</li>
<li>XGL: dev 87430a6fa88da11473eed49603def2fa47b51fa0</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 88613f67ca2e3e64cc1c7c4ccd47e126495e6e23</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/143766)</summary><ul><li>Built with version 1.2.2.2</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 26276/47929 (54.8%)</li>
<li>Failed: 0/47929 (0.0%)</li>
<li>Skipped: 21653/47929 (45.2%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 26275/47929 (54.8%)</li>
<li>Failed: 0/47929 (0.0%)</li>
<li>Skipped: 21654/47929 (45.2%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 24807/47908 (51.8%)</li>
<li>Failed: 0/47908 (0.0%)</li>
<li>Skipped: 23101/47908 (48.2%)</li>
</ul></details>


</ul></details>

**Comment by kuhar** [comment]
Ah, I see what you mean. In other words, we know that `points-to(const)` is an empty set because we don't have any magic constants representing special memory locations.

Yeah, in this case this transformation seems safe to me. Could you please update the code and mention why LLVM doesn't optimize this but we can?

**Comment by rdomingu** [comment]
The last push was to squash the two commits into one. No functional or any other changes. I will be making another push with the comments regarding this transformation in LLPC vs LLVM and the LPC lit test.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 7d9b87555c57c65b2a39dd3715cd4dd350685c07
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 4ae736bdbc5d5dee59851ac564c5e21d807b44b0</li>
<li>SPVGEN: dev e1f1881d84cd1c7a6c1d661cbac0e619b4e110c1</li>
<li>XGL: dev 87430a6fa88da11473eed49603def2fa47b51fa0</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 88613f67ca2e3e64cc1c7c4ccd47e126495e6e23</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/143766)</summary><ul><li>Built with version 1.2.2.2</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 26276/47929 (54.8%)</li>
<li>Failed: 0/47929 (0.0%)</li>
<li>Skipped: 21653/47929 (45.2%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 26275/47929 (54.8%)</li>
<li>Failed: 0/47929 (0.0%)</li>
<li>Skipped: 21654/47929 (45.2%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 24807/47908 (51.8%)</li>
<li>Failed: 0/47908 (0.0%)</li>
<li>Skipped: 23101/47908 (48.2%)</li>
</ul></details>


</ul></details>

**Comment by amdvlk-admin** [comment]
### Test summary for commit 44778722256c029e12383d3f0d3ef5f6d693877f
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 4ae736bdbc5d5dee59851ac564c5e21d807b44b0</li>
<li>SPVGEN: dev e1f1881d84cd1c7a6c1d661cbac0e619b4e110c1</li>
<li>XGL: dev 87430a6fa88da11473eed49603def2fa47b51fa0</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 88613f67ca2e3e64cc1c7c4ccd47e126495e6e23</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/143766)</summary><ul><li>Built with version 1.2.2.2</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 26276/47929 (54.8%)</li>
<li>Failed: 0/47929 (0.0%)</li>
<li>Skipped: 21653/47929 (45.2%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 26275/47929 (54.8%)</li>
<li>Failed: 0/47929 (0.0%)</li>
<li>Skipped: 21654/47929 (45.2%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 24807/47908 (51.8%)</li>
<li>Failed: 0/47908 (0.0%)</li>
<li>Skipped: 23101/47908 (48.2%)</li>
</ul></details>


</ul></details>

**Review by kuhar — APPROVED** [review]

File: lgc/patch/PatchPeepholeOpt.cpp [hunk 2]
```diff
@@ -571,6 +571,53 @@ void PatchPeepholeOpt::visitPHINode(PHINode &phiNode) {
...
+  const ConstantInt *const offset = dyn_cast<ConstantInt>(binaryOperator->getOperand(1));
+  if (!offset)
+    return;
+
+  // Change inttoptr ( add x, const ) -> gep ( inttoptr x, const )
+  IntToPtrInst *const newIntToPtr = new IntToPtrInst(binaryOperator->getOperand(0), intToPtr.getType());
+  insertAfter(*newIntToPtr, *binaryOperator);
+
+  // Create a getelementptr instruction (using offset / size).
+  const DataLayout DL = newIntToPtr->getModule()->getDataLayout();
```
Comment (amdrexu) [inline lgc/patch/PatchPeepholeOpt.cpp:604]: DL -> dataLayout

File: lgc/patch/PatchPeepholeOpt.cpp [hunk 3]
```diff
@@ -571,6 +571,53 @@ void PatchPeepholeOpt::visitPHINode(PHINode &phiNode) {
...
+//
+// @param intToPtr: The "inttoptr" instruction to visit.
+void PatchPeepholeOpt::visitIntToPtr(IntToPtrInst &intToPtr) {
+  // Check if we are using add to do pointer arithmetic.
+  BinaryOperator *const binaryOperator = dyn_cast<BinaryOperator>(intToPtr.getOperand(0));
+  if (!binaryOperator || binaryOperator->getOpcode() != Instruction::Add)
+    return;
+
+  // Check that we have a constant offset.
+  const ConstantInt *const offset = dyn_cast<ConstantInt>(binaryOperator->getOperand(1));
```
Comment (amdrexu) [inline lgc/patch/PatchPeepholeOpt.cpp:595]: offset -> constOffset

**Review by amdrexu — COMMENTED** [review]
Some nits.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 2d9d1dd8788458e7e46a2342cad7d86f39b481d8
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 4ae736bdbc5d5dee59851ac564c5e21d807b44b0</li>
<li>SPVGEN: dev e1f1881d84cd1c7a6c1d661cbac0e619b4e110c1</li>
<li>XGL: dev 87430a6fa88da11473eed49603def2fa47b51fa0</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 88613f67ca2e3e64cc1c7c4ccd47e126495e6e23</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/144366)</summary><ul><li>Built with version 1.2.2.2</li>
<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 26276/48129 (54.6%)</li>
<li>Failed: 0/48129 (0.0%)</li>
<li>Skipped: 21853/48129 (45.4%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 26275/48129 (54.6%)</li>
<li>Failed: 0/48129 (0.0%)</li>
<li>Skipped: 21854/48129 (45.4%)</li>
</ul></details>

<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 24807/48108 (51.6%)</li>
<li>Failed: 0/48108 (0.0%)</li>
<li>Skipped: 23301/48108 (48.4%)</li>
</ul></details>


</ul></details>

**Review by trenouf — COMMENTED** [review]
Sorry I'm a bit late to this review party, but why is this necessary? Is the spir-v reader generating add then i2p? Or has some earlier optimization generated them?

**Comment by rdomingu** [comment]
This is coming from the SPIR-V producer.

**Comment by trenouf** [comment]
Could we fix that instead?

**Comment by rdomingu** [comment]
This is out of our control. I can't change how the SPIR-V producer manipulates the pointer.

**Comment by trenouf** [comment]
Oh, sorry, I misread what you said as "spir-v reader" (i.e. the thing that generates the llvm IR).

**Review by amdrexu — APPROVED** [review]

**Pull request merged**